backend/views/retrieve_test_view.py

```diff
@@ -79,85 +101,172 @@ def retrieve_json_from_name_date(test_name, query_date_time, request_type):
-            fr_id = Language.objects.get(
-                ISO_Code_2="fr-ca", date_from__lte=query_date_time, date_to__isnull=True
-            ).language_id
-        except Language.DoesNotExist:
-            fr_id = None
+    en_id = get_language("en-ca", query_date_time)
+    fr_id = get_language("fr-ca", query_date_time)
     return en_id, fr_id
 
 
+def get_language(iso_code_2, query_date_time):
+    # get the active language id for the given iso_code_2
+    lang_id = Language.objects.filter(ISO_Code_2=iso_code_2)
+    lang_id = exclude_inactive_objects(lang_id, query_date_time)
+    if not lang_id:
+        return None
+    return lang_id.last()
+
+
 def get_text_detail(item_id, language_id, query_date_time):
-    try:
-        text_detail = ItemText.objects.get(
-            item_id=item_id,
-            language=language_id,
-            date_from__lte=query_date_time,
-            date_to__gt=query_date_time,
-        ).text_detail
-    except ItemText.DoesNotExist:
-        try:
-            text_detail = ItemText.objects.get(
-                item_id=item_id,
-                language=language_id,
-                date_from__lte=query_date_time,
-                date_to__isnull=True,
-            ).text_detail
-        except ItemText.DoesNotExist:
-            text_detail = None
-    return text_detail
+    # get the itemtext object and text_detail for the given item_id and language
+    item_text = ItemText.objects.filter(
+        item_id=item_id,
+        language=language_id
+    )
+    item_text = exclude_inactive_objects(item_text, query_date_time)
+    if not item_text:
+        # if empty
+        return None
+    return item_text.last().text_detail
 
 
 def get_item_by_id(item_id, query_date_time):
-    try:
-        item = Item.objects.get(
-            pk=item_id, date_from__lte=query_date_time, date_to__gt=query_date_time
-        )
-    except Item.DoesNotExist:
-        # if the above fails, try to get an item with the same id, after the from date
-        # and where the to date is null
-        try:
-            item = Item.objects.get(
-                pk=item_id, date_from__lte=query_date_time, date_to__isnull=True
-            )
-        except Item.DoesNotExist:
-            item = None
-    return item
+    # get the item by id
+    item = Item.objects.filter(pk=item_id)
+    item = exclude_inactive_objects(item, query_date_time)
+    if not item:
+        # if empty
+        return None
+    return item.last()
+
+
+def get_items_by_parent_id(parent_id, query_date_time):
+    # get all items with the given parent_id
+    items = Item.objects.filter(parent_id=parent_id)
+    items = exclude_inactive_objects(items, query_date_time)
+    return items
+
+
+def exclude_inactive_objects(objects, query_date_time):
+    # exclude all with later date_from or earlier date_to
+    # however, exactly equal are acceptible
+    return objects.exclude(date_from__gt=query_date_time).exclude(date_to__lt=query_date_time)

```

Comment:
It's possible that this causes performance issues because it looks like this is done in python and wouldn't be optimized in SQL.

Reply:
according to https://docs.djangoproject.com/en/2.2/topics/db/queries/#querysets-are-lazy query sets are not actually evaluated till you "ask for them", so in this case it is fine, however, I will look through the code and see if there are any obvious issues.